channels/cliprdr/client/cliprdr_format.c

```diff
@@ -336,3 +336,181 @@ UINT cliprdr_process_format_data_response(cliprdrPlugin* cliprdr, wStream* s, UI
+			count * CLIPRDR_FILEDESCRIPTOR_SIZE,
+			Stream_GetRemainingLength(s));
+
+		result = ERROR_INCORRECT_SIZE;
+		goto out;
+	}
+
+	*file_descriptor_count = count;
+	*file_descriptor_array = calloc(count, sizeof(FILEDESCRIPTOR));
+	if (!*file_descriptor_array)
+	{
+		result = ERROR_NOT_ENOUGH_MEMORY;
+		goto out;
+	}
+
+	for (i = 0; i < count; i++)
+	{
+		int c;
+		UINT64 lastWriteTime;
+		FILEDESCRIPTOR* file = &((*file_descriptor_array)[i]);
+
+		Stream_Read_UINT32(s, file->dwFlags); /* flags (4 bytes) */
+		Stream_Seek(s, 32); /* reserved1 (32 bytes) */
+		Stream_Read_UINT32(s, file->dwFileAttributes); /* fileAttributes (4 bytes) */
+		Stream_Seek(s, 16); /* reserved2 (16 bytes) */
+		Stream_Read_UINT64(s, lastWriteTime); /* lastWriteTime (8 bytes) */
+		file->ftLastWriteTime = uint64_to_filetime(lastWriteTime);
+		Stream_Read_UINT32(s, file->nFileSizeHigh); /* fileSizeHigh (4 bytes) */
+		Stream_Read_UINT32(s, file->nFileSizeLow); /* fileSizeLow (4 bytes) */
+		for (c = 0; c < 260; c++) /* cFileName (520 bytes) */
+			Stream_Read_UINT16(s, file->cFileName[c]);
+	}
+
+	if (Stream_GetRemainingLength(s) > 0)
+		WLog_WARN(TAG, "packed file list has %"PRIuz" excess bytes",
+			Stream_GetRemainingLength(s));
+out:
+	Stream_Free(s, FALSE);
+
+	return result;
+}
+
+#define CLIPRDR_MAX_FILE_SIZE (2U * 1024 * 1024 * 1024)
+
+/**
+ * Serialize a packed file list.
+ *
+ * The resulting format data must be freed with the `free()` function.
+ *
+ * @param [in]  file_descriptor_array  array of `FILEDESCRIPTOR` structs to serialize.
+ * @param [in]  file_descriptor_count  number of elements in `file_descriptor_array`.
+ * @param [out] format_data            serialized CLIPRDR_FILELIST.
+ * @param [out] format_data_length     length of `format_data` in bytes.
+ *
+ * @returns 0 on success, otherwise a Win32 error code.
+ */
+UINT cliprdr_serialize_file_list(const FILEDESCRIPTOR* file_descriptor_array,
+		UINT32 file_descriptor_count, BYTE** format_data, UINT32* format_data_length)
+{
+	UINT result = NO_ERROR;
+	UINT32 i;
+	wStream* s = NULL;
+
+	if (!file_descriptor_array || !format_data || !format_data_length)
+		return ERROR_BAD_ARGUMENTS;
+
+	s = Stream_New(NULL, 4 + file_descriptor_count * CLIPRDR_FILEDESCRIPTOR_SIZE);
+	if (!s)
+		return ERROR_NOT_ENOUGH_MEMORY;
+
+	Stream_Write_UINT32(s, file_descriptor_count); /* cItems (4 bytes) */
+
+	for (i = 0; i < file_descriptor_count; i++)
+	{
+		int c;
+		UINT64 lastWriteTime;
+		const FILEDESCRIPTOR* file = &file_descriptor_array[i];
+
+		/*
+		 * There is a known issue with Windows server getting stuck in
+		 * an infinite loop when downloading files that are larger than
+		 * 2 gigabytes. Do not allow clients to send such file lists.
+		 *
+		 * https://support.microsoft.com/en-us/help/2258090
+		 */
+		if ((file->nFileSizeHigh > 0) || (file->nFileSizeLow >= CLIPRDR_MAX_FILE_SIZE))
+		{
+			WLog_ERR(TAG, "cliprdr does not support files over 2 GB");
+			result = ERROR_FILE_TOO_LARGE;
```

Comment:
Maybe. I put it there because I guessed that all clients will have to perform such checks if they want to deal with that issue, so why not put this failsafe in the common function. However, it still requires cooperation from the client. If the client does not fill in the file size immediately in FILEDESCRIPTOR list then the server can still break later when it explicitly requests the file size and proceeds with content retrieval (as there are not checks there).

I placed that check before any file content transfer in order to abort the whole clipboard paste if there are files like that. If we fail it later then remote Windows will just stop, show no error messages, and paste random files which it has managed to download so far. I believe this is wrong user experience as it suggests successful transfer for the user. But it is mainly caused by the fact that xfreerdp does not show any error messages. I guess there would be no problem with aborting the transfer halfway if we (more prominently) tell the user what's broke.

Also, I did not put these checks into wClipboard stuff because I felt that a 'generic clipboard' should not deal with peculiarities of cliprdr protocol and its server-side implementation. However, it may be a good place for them, as the final frontier where we can prevent a disaster on the server.

Another option will be to remove this failsafe entirely. I do not remember which systems actually fail catastrophically if they are faced with 2GB+ files. Maybe it's only the oldest ones.

I have no problem with moving this check into the client just before calling cliprdr_serialize_file_list(), just don't see any practical merit in it at the moment.

@akallabeth, what's your opinon on this?